src/System.Data.Common/src/System/Data/Common/DbCommand.cs

```diff
@@ -0,0 +1,265 @@
+        }
+
+
+        abstract protected DbParameter CreateDbParameter();
+
+        abstract protected DbDataReader ExecuteDbDataReader(CommandBehavior behavior);
+
+        abstract public int ExecuteNonQuery();
+
+        public DbDataReader ExecuteReader()
+        {
+            return (DbDataReader)ExecuteDbDataReader(CommandBehavior.Default);
+        }
+
+
+        public DbDataReader ExecuteReader(CommandBehavior behavior)
+        {
+            return (DbDataReader)ExecuteDbDataReader(behavior);
+        }
+
+
+        public Task<int> ExecuteNonQueryAsync()
+        {
+            return ExecuteNonQueryAsync(CancellationToken.None);
+        }
+
+        public virtual Task<int> ExecuteNonQueryAsync(CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return ADP.CreatedTaskWithCancellation<int>();
+            }
+            else
+            {
+                CancellationTokenRegistration registration = new CancellationTokenRegistration();
+                if (cancellationToken.CanBeCanceled)
+                {
+                    registration = cancellationToken.Register(CancelIgnoreFailure);
+                }
+
+                try
+                {
+                    return Task.FromResult<int>(ExecuteNonQuery());
+                }
+                catch (Exception e)
+                {
+                    registration.Dispose();
+                    return ADP.CreatedTaskWithException<int>(e);
+                }
+            }
+        }
+
+        public Task<DbDataReader> ExecuteReaderAsync()
+        {
+            return ExecuteReaderAsync(CommandBehavior.Default, CancellationToken.None);
+        }
+
+        public Task<DbDataReader> ExecuteReaderAsync(CancellationToken cancellationToken)
+        {
+            return ExecuteReaderAsync(CommandBehavior.Default, cancellationToken);
+        }
+
+        public Task<DbDataReader> ExecuteReaderAsync(CommandBehavior behavior)
+        {
+            return ExecuteReaderAsync(behavior, CancellationToken.None);
+        }
+
+        public Task<DbDataReader> ExecuteReaderAsync(CommandBehavior behavior, CancellationToken cancellationToken)
+        {
+            return ExecuteDbDataReaderAsync(behavior, cancellationToken);
+        }
+
+        protected virtual Task<DbDataReader> ExecuteDbDataReaderAsync(CommandBehavior behavior, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return ADP.CreatedTaskWithCancellation<DbDataReader>();
+            }
+            else
+            {
+                CancellationTokenRegistration registration = new CancellationTokenRegistration();
+                if (cancellationToken.CanBeCanceled)
+                {
+                    registration = cancellationToken.Register(CancelIgnoreFailure);
+                }
+
+                try
```

Comment:
Same comments as in the previous function.  I'll refrain from commenting on these further if they arise again.